components/learn/modules/ROOT/pages/on-upgrades.adoc

```diff
@@ -1,8 +1,281 @@
+
+$ npx openzeppelin call
+? Pick a network: development
+? Pick an instance: Box at 0xCfEB869F69431e42cdB54A4F4f105C19C080A601
+? Select which function: retrieve()
+✓ Method 'retrieve()' returned: 6
+```
+
+That's it! Notice how the `value` of the `Box` was preserved throughout the upgrade, as well as its address. And this process is the same regardless of whether you are working on a local blockchain, a testnet, or the main network. Let's see how the OpenZeppelin SDK accomplishes this.
+
+[[how-upgrades-work]]
+== How upgrades work
+
+_This section will be more theory-heavy than others, so feel free to skip over it, and return later if you are curious._
+
+When you create a new upgradeble contract instance, the OpenZeppelin SDK actually deploys two contracts:
+
+. The contract you have written, which is known as the _implementation contract_ or _logic contract_.
+. A _proxy_ to that contract, which is the instance that you actually interact with.
+
+Here, the _proxy_ is a simple contract that just _delegates_ all calls to an implementation contract. A _delegate call_ is similar to a regular call, but all code is executed in the context of the caller, not of the callee. This means that a `transfer` in the implementation contract's code will actually transfer the proxy's balance. Similarly, any reads or writes to the contract storage will read or write from the proxy's storage.
+
+This allows us to **decouple** a contract's state and code: the proxy holds the state, while the logic contract provides the code. And it also allows us to **change** the code by just having the proxy delegate to a different implementation contract.
+
+An upgrade then involves the following steps:
+
+. Deploy the new implementation contract.
+. Send a transaction to the proxy, that updates its implementation address to the new one.
+
+NOTE: You can have multiple proxies using the same implementation contract, so you can save gas using this pattern if you plan to deploy multiple copies of the same contract.
+
+Any user of the smart contract always interacts with the proxy, which never changes its address. This allows you to roll out an upgrade or fix a bug without requesting your users to change anything on their end - they just keep interacting with the same address as always.
+
+NOTE: If you want to learn more about how OpenZeppelin proxies work, check out our xref:sdk::pattern.adoc[Upgrades Pattern guide].
+
+[[limitations-of-contract-upgrades]]
+== Limitations of Contract Upgrades
+
+While any smart contract can be made upgradeable, some restrictions of the Solidity language need to be worked around. These come up when writing both the initial contract and version we'll upgrade it to.
+
+=== Initialization
+
+When writing an upgradeable contract, you cannot declare a `constructor`. If you want to run any initialization code, the SDK provides a special `Initializable` base contract that allows you to tag a method as `initializer`, ensuring it can be run only once.
```

Comment:
```suggestion
Upgradeable contracts cannot have a `constructor`. To help you run initialization code, the SDK provides the `Initializable` base contract that allows you to tag a method as `initializer`, ensuring it can be run only once.
```